Create react best practises wiki (#64)
amdigital-co-uk/docs-playbook: docs/Delivery-Practices/Build-and-Release/Engineering/Quality-Standards/React-Best-Practises.md

---
title: React Best Practises 
authors: 
- Rhodri Hewitson
---

## Introduction

React is a popular open-source JavaScript library used for building user interfaces. It was created by Facebook and has gained a lot of popularity among developers for its declarative and efficient approach to building web applications. TypeScript is a superset of JavaScript that adds optional static typing and other features to the language. By using TypeScript with React, developers can write more reliable, maintainable, and scalable code with fewer errors. In this context, it's important to follow best practices and guidelines to ensure that your code is of the highest quality. In this post, we've provided some guidelines and best practices for writing React code in TypeScript that can help you write better code and improve the overall quality of your application.

## Guidelines for Writing React Code in TypeScript

> 1. Use semantic HTML: Use semantic HTML tags that describe the content of the page rather than just the appearance. This will help with accessibility and search engine optimization, and make your code easier to read and maintain.

Reviewer: It may worth adding the relevant documentation for these guidelines. Something like: "There's a lot of documentation out there for semantic HTML, an example of a popular site used for web standards is [MDN](https://developer.mozilla.org/en-US/docs/Glossary/Semantics).

1. Use React Query: React Query is a library for managing data fetching and caching in React applications. It provides a simple and intuitive API for fetching data from APIs, and supports caching, pagination, and other advanced features.

1. Use strict mode: Use the React strict mode to enable additional checks and warnings for potential problems in your code. This can help you catch potential issues early and improve the overall quality of your code.

1. Use ESLint and Prettier: Use ESLint and Prettier to enforce coding standards and formatting rules in your codebase. These tools can help you maintain consistency and improve the readability of your code.

1. Use CSS modules or styled-components: Use CSS modules or styled-components to manage the styling of your components. These tools provide a way to write scoped styles that are only applied to specific components and can help you avoid naming collisions and other styling issues.

> 1. Use prop drilling sparingly: Avoid excessive prop drilling by using React Query or React Context. These provide a centralized store for managing the state of your application and can help you avoid passing props down through multiple levels of components.

Reviewer: Prop drilling isn't ideal at all, but using React Context can have issues of its own. Essentially, when a context value is changed, all child components will re-render. This loses the value in the virtual DOM, so context should only be used if a value isn't going to be changed frequently and it's value is going to be used in multiple components

E.g. https://blog.logrocket.com/pitfalls-of-overusing-react-context/

1. Use async/await for data fetching: Use async/await syntax to fetch data asynchronously from APIs or other sources. This can help you avoid callback hell and make your code easier to read and maintain.

> 1. Use hooks effectively: Use React hooks like useState and useEffect to manage state and side effects in your components. Hooks provide a simple and intuitive API for managing state and can help you avoid class components and the complexities that come with them.

Reviewer: I'd also highlight to use them as sparingly as possible. 

1. useState
  - Ensure we are using "derived state" wherever possible and not "state duplication"
    - https://blog.webdevsimplified.com/2019-11/never-store-derived-state/
  - useState causes the DOM to rerender. Only use useState if we want rerenders to occur. If not, you may want `useRef`

1. useEffect
  - Used for syncing data; however, this can also cause a derived state issue again

## 

## Guidelines for Writing React Code in TypeScript

1. Use semantic HTML: Use semantic HTML tags that describe the content of the page rather than just the appearance. This will help with accessibility and search engine optimization, and make your code easier to read and maintain.

1. Use React Testing Library: Use React Testing Library to write unit and integration tests for your components. This library provides a simple and intuitive API for testing React components and encourages developers to write tests that reflect the way users interact with the application.

1. Use the Atomic Design Methodology: The Atomic Design Methodology is a system for designing and building UI components in a structured and scalable way. By breaking down your UI into smaller, reusable components, you can create a more flexible and maintainable codebase.

> 1. Use React Query: React Query is a library for managing data fetching and caching in React applications. It provides a simple and intuitive API for fetching data from APIs, and supports caching, pagination, and other advanced features.

Reviewer: Not sure it's worth highlighting to not use `onSuccess`/`onError`/`onSettled` when using `useQuery` since it will be deprecated in future. 

https://twitter.com/TkDodo/status/1647341498227097600

1. Use strict mode: Use the React strict mode to enable additional checks and warnings for potential problems in your code. This can help you catch potential issues early and improve the overall quality of your code.

1. Use ESLint and Prettier: Use ESLint and Prettier to enforce coding standards and formatting rules in your codebase. These tools can help you maintain consistency and improve the readability of your code.

1. Use CSS modules or styled-components: Use CSS modules or styled-components to manage the styling of your components. These tools provide a way to write scoped styles that are only applied to specific components and can help you avoid naming collisions and other styling issues.

1. Use prop drilling sparingly: Avoid excessive prop drilling by using React Query or React Context. These provide a centralized store for managing the state of your application and can help you avoid passing props down through multiple levels of components.

1. Use async/await for data fetching: Use async/await syntax to fetch data asynchronously from APIs or other sources. This can help you avoid callback hell and make your code easier to read and maintain.

> 1. Use hooks effectively: Use React hooks like useState and useEffect to manage state and side effects in your components. Hooks provide a simple and intuitive API for managing state and can help you avoid class components and the complexities that come with them. But remember [you may not always need an effect](https://react.dev/learn/you-might-not-need-an-effect#)

Reviewer: Also, prevent using objects in your `useEffect`: https://stackoverflow.com/questions/54095994/react-useeffect-comparing-objects

---
title: React Best Practises 
authors: 
- Rhodri Hewitson
---

## Introduction

React is a popular open-source JavaScript library used for building user interfaces. It was created by Facebook and has gained a lot of popularity among developers for its declarative and efficient approach to building web applications. TypeScript is a superset of JavaScript that adds optional static typing and other features to the language. By using TypeScript with React, developers can write more reliable, maintainable, and scalable code with fewer errors. In this context, it's important to follow best practices and guidelines to ensure that your code is of the highest quality. In this post, we've provided some guidelines and best practices for writing React code in TypeScript that can help you write better code and improve the overall quality of your application.

> ## 

Reviewer: Does this need to be removed?

1. Use React Testing Library: Use React Testing Library to write unit and integration tests for your components. This library provides a simple and intuitive API for testing React components and encourages developers to write tests that reflect the way users interact with the application. See [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) for more information.

1. Use the Atomic Design Methodology: The Atomic Design Methodology is a system for designing and building UI components in a structured and scalable way. By breaking down your UI into smaller, reusable components, you can create a more flexible and maintainable codebase. See Brad Frost article on [Atomic Design](https://atomicdesign.bradfrost.com/table-of-contents/) for more information.

1. Use React Query: React Query is a library for managing data fetching and caching in React applications. It provides a simple and intuitive API for fetching data from APIs, and supports caching, pagination, and other advanced features. See [ReactQuery](https://tanstack.com/query/v3/docs/react/overview) for more infomation.

> _Note `onSuccess()/onError()/onSettled()` will be deprecated see [React Query v5](https://twitter.com/TkDodo/status/1647341498227097600)._

Reviewer: This is only being deprecated in `useQuery`, it'll still be available in `useMutation`

1. Read and understand the official React documentation: The React documentation at <https://react.dev/> is a valuable resource for learning about React and its ecosystem. Take the time to read and understand the documentation, and follow the best practices and recommendations provided by the React team. This will help you write better code and avoid common pitfalls in your React applications.

> 1. Use strict mode: Use the React strict mode to enable additional checks and warnings for potential problems in your code. This can help you catch potential issues early and improve the overall quality of your code.

Reviewer: ```suggestion
1. Use strict mode: Use the [React strict mode](https://react.dev/reference/react/StrictMode) to enable additional checks and warnings for potential problems in your code. This can help you catch potential issues early and improve the overall quality of your code.
```

> 1. Use ESLint and Prettier: Use ESLint and Prettier to enforce coding standards and formatting rules in your codebase. These tools can help you maintain consistency and improve the readability of your code.

Reviewer: ```suggestion
1. Use ESLint and Prettier: Use [ESLint](https://eslint.org/) and [Prettier](https://prettier.io/) to enforce coding standards and formatting rules in your codebase. These tools can help you maintain consistency and improve the readability of your code.
```

1. Use CSS modules or styled-components: Use CSS modules or styled-components to manage the styling of your components. These tools provide a way to write scoped styles that are only applied to specific components and can help you avoid naming collisions and other styling issues.

1. Be careful of excessive prop drilling: Avoid excessive prop drilling by using React Query or React Context. These provide a centralized store for managing the state of your application and can help you avoid passing props down through multiple levels of components. Alternatively lifting the state up can help reduce prop drilling see Kent C Dodds article on [application state management with react](https://kentcdodds.com/blog/application-state-management-with-react). React.dev also has a great tutorial on [managing state](https://react.dev/learn/managing-state). _Note we haven't yet needed a global state but lightweight libraries like [zustand](https://github.com/pmndrs/zustand) or [jotai](https://github.com/pmndrs/jotai) could fit that need depending on the use case._

> 1. Use async/await for data fetching: Use async/await syntax to fetch data asynchronously from APIs or other sources. This can help you avoid callback hell and make your code easier to read and maintain.

Reviewer: ```suggestion
1. Use async/await for data fetching: Use async/await syntax to fetch data asynchronously from APIs or other sources as opposed to using Promises. Using async/await syntax can help you avoid callback hell, which can be brought on by using Promises, and also make your code easier to read and maintain.
```
